### ISSUE TITLE
`actionFinally` handlers can be incorrectly run twice on ctrl-C

### ISSUE THREAD

**Issue by judah** [ndmitchell/shake#611]
When you Ctrl-C a Shake program, it's possible for pending `actionFinally` commands to be run *twice* by Shake.  This appears to be a race condition in Shake's cleanup code.

I discovered this in judah/pier#100.  Pier uses `actionFinally` to clean up temporary directories with  Occasionally, when you control-C `pier`, you get an error message indicating that two `removeDirectoryRecursive` calls on the same temporary directory are running simultaneously.

Example code to reproduce the failure: 
```
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeFamilies #-}
module Main (main) where

import Development.Shake
import Development.Shake.Classes
import Development.Shake.Rule
import GHC.Generics
import Control.Concurrent

main :: IO ()
main = shakeArgs shakeOptions $ do
    r <- addOracle $ \(MyRule k) -> do
        putNormal $ "Running " ++ show k
        let delay = do
                    putStrLn $ "Delay cleanup " ++ show k
                    threadDelay 1000000
        let finish = do
                    putStrLn $ "Finish cleanup " ++ show k
                    threadDelay 1000000
        liftIO (threadDelay 3000000)
            `actionFinally` delay `actionFinally` finish
    action $ forP [1..4] $ \k -> r (MyRule k) >>= putNormal . show

data MyRule = MyRule Int
    deriving (Show, Typeable, Eq, Generic, Hashable, Binary, NFData)
type instance RuleResult MyRule = ()
```

Then, run:
```
$ stack --resolver=lts-12.8 exec ghc -- -threaded Reproduce.hs
$ ./Reproduce -j2
```
And press ctrl-C while it's still running.

The output that I got was:
```
Running 1
Running 3
^CDelay cleanup 3
Delay cleanup 1
Finish cleanup 3
Finish cleanup 1
Finish cleanup 3
Finish cleanup 1
user interrupt
```

That the messages "Finish cleanup 1" and "Finish cleanup 3" are printed *twice*, even though the corresponding "Running" commands are only printed once.

**Comment by judah** [comment]
Here's why I think the bug happens.  This is the code for `actionBoom`:
```
actionBoom :: Bool -> Action a -> IO b -> Action a
actionBoom runOnSuccess act clean = do
    Global{..} <- Action getRO
    undo <- liftIO $ addCleanup globalCleanup $ void clean
    -- important to mask_ the undo/clean combo so either both happen or neither
    res <- Action $ catchRAW (fromAction act) $ \e -> liftIO (mask_ $ undo >> clean) >> throwRAW e
    liftIO $ mask_ $ undo >> when runOnSuccess (void clean)
    return res
```

When you press Ctrl-C, Shake handles it by:
1. Sending a `ThreadKilled` exception to each thread (but not waiting for them to finish).
2. Shake's main thread then runs every cleanup which hasn't had `undo` run for it already.
3. Simultaneously, each thread then runs their `catchRAW` handler, which calls `undo` (without checking whether it was already run in (2) and then runs the cleanup unconditionally.

So it's possible for a given `actionFinally` handler to run in both steps (2) and (3).

**Comment by ndmitchell** [comment]
Thanks for the report, and thanks for the diagnosis - I fixed it by having catchRAW check whether undo actually removed the cleanup from the map (and thus should execute the cleanup), or if it had already been removed. Test added.

How problematic is this bug? I expect to release a new version of Shake in the next 2 weeks, but if it's particularly annoying, I can do so sooner. (I'd normally release immediately for a bug of this magnitude, but the repo is in enough flux that I'd wait if I could)
